lib/zoneinfo_module.c

```diff
@@ -0,0 +1,2191 @@
+    .tp_new = zoneinfo_new,
+    .tp_dealloc = zoneinfo_dealloc,
+};
+
+/////
+// Specify the zoneinfo._czoneinfo module
+static PyMethodDef module_methods[] = {{NULL, NULL}};
+static void
+module_free()
+{
+    Py_XDECREF(TIMEDELTA_CACHE);
+    TIMEDELTA_CACHE = NULL;
+
+    Py_XDECREF(ZONEINFO_WEAK_CACHE);
+    ZONEINFO_WEAK_CACHE = NULL;
+
+    Py_XDECREF(_tzpath_find_tzfile);
+    _tzpath_find_tzfile = NULL;
+
+    Py_XDECREF(_common_mod);
+    _common_mod = NULL;
+
+    Py_XDECREF(io_open);
+    io_open = NULL;
+}
+
+static struct PyModuleDef zoneinfomodule = {
+    PyModuleDef_HEAD_INIT,
+    .m_name = "zoneinfo._czoneinfo",
+    .m_doc = "C implementation of the zoneinfo module",
+    .m_size = -1,
+    .m_methods = module_methods,
+    .m_free = module_free};
+
+PyMODINIT_FUNC
+PyInit__czoneinfo(void)
+{
+    PyObject *m; /* a module object */
+    m = PyModule_Create(&zoneinfomodule);
+
+    if (m == NULL) {
+        return NULL;
+    }
+
+    PyDateTime_IMPORT;
+    PyZoneInfo_ZoneInfoType.tp_base = PyDateTimeAPI->TZInfoType;
+    if (PyType_Ready(&PyZoneInfo_ZoneInfoType) < 0) {
+        return NULL;
+    }
+
+    Py_INCREF(&PyZoneInfo_ZoneInfoType);
+    PyModule_AddObject(m, "ZoneInfo", (PyObject *)&PyZoneInfo_ZoneInfoType);
+
+    /* Populate imports */
+    PyObject *_tzpath_module = PyImport_ImportModule("zoneinfo._tzpath");
+    if (_tzpath_module == NULL) {
+        return NULL;
+    }
+
+    _tzpath_find_tzfile =
+        PyObject_GetAttrString(_tzpath_module, "find_tzfile");
+    Py_DECREF(_tzpath_module);
+    if (_tzpath_find_tzfile == NULL) {
+        return NULL;
+    }
+
+    PyObject *io_module = PyImport_ImportModule("io");
+    if (io_module == NULL) {
+        return NULL;
+    }
+
+    io_open = PyObject_GetAttrString(io_module, "open");
+    Py_DECREF(io_module);
+    if (io_open == NULL) {
+        return NULL;
+    }
+
+    _common_mod = PyImport_ImportModule("zoneinfo._common");
+    if (_common_mod == NULL) {
+        return NULL;
+    }
+
+    /* Initialize caches */
+    TIMEDELTA_CACHE = PyDict_New();
+
+    PyObject *weakref_module = NULL;
+    PyObject *WeakValueDictionary = NULL;
+
+    if ((weakref_module = PyImport_ImportModule("weakref")) == NULL) {
+        return NULL;
+    }
+
+    WeakValueDictionary =
+        PyObject_GetAttrString(weakref_module, "WeakValueDictionary");
+    Py_DECREF(weakref_module);
+    weakref_module = NULL;
+    if (WeakValueDictionary == NULL) {
+        return NULL;
+    }
+    PyObject *no_args = PyTuple_New(0);
+    if (no_args == NULL) {  // GCOV_EXCL_START
+        Py_DECREF(WeakValueDictionary);
+        return NULL;
+    }  // GCOV_EXCL_STOP
+
+    // TODO: VectorCall
+    ZONEINFO_WEAK_CACHE = PyObject_CallObject(WeakValueDictionary, no_args);
+
+    Py_DECREF(no_args);
+    Py_DECREF(WeakValueDictionary);
+    if (ZONEINFO_WEAK_CACHE == NULL) {  // GCOV_EXCL_START
+        return NULL;
+    }  // GCOV_EXCL_STOP
+
+    return m;
+}
```

Comment:
Does `TIMEDELTA_CACHE` need a decref in the error conditions or would that be handled elsewhere? If so, the `goto error` pattern might be useful here.

It seems like there could also be an error condition to check if `TIMEDELTA_CACHE` couldn't be allocated. I.E.:

```c
     TIMEDELTA_CACHE = PyDict_New();
     if (TIMEDELTA_CACHE == NULL) {
         return NULL;
     }
```

(I'm not certain about the above, it just caught my attention)

Reply:
I think you're right that we should return `NULL` if `TIMEDELTA_CACHE` fails to initialize, but I believe that `module_free` will be executed even if there's an error during initialization, so the `Py_XDECREF` in the free function should handle cleanup for us.

Reply:
Ah okay, I wasn't aware that `module_free` would still be executed in that situation. This is my first in-depth review of an entire new C extension module, so there's a decent amount that I'm unaware of. But, `Py_XDECREF` does seem correct over `Py_DECREF`, since `TIMEDELTA_CACHE` could potentially be `NULL`.


P.S. So far, I've spent 90% of my contribution time for CPython on committing and reviewing pure Python code in the stdlib (mainly asyncio and concurrent.futures) and documentation. I'm certainly interested in learning more about the C-API and C internals though, so I figured that this would be a good opportunity while providing some help at the same time. :-)

But if you told me ~a year ago when I opened my first PR that I'd be helping to review a new timezone C extension module in the near future, I probably wouldn't believe you ;)